File: patches/kconfig/kde2elektra.php
#!/usr/bin/php 
<?
/*
 * KDE Configuration import
 *
 * Yannick Lecaillez
 * <sizon5@gmail.com>
 */

function convert($file, $rootKey)
{
	$currentGroup = "<default>";
	$keyName = NULL;
	$keyValue = NULL;

/*	if ( substr($file, -2) != 'rc' )
		return; */
	
	$fp = fopen($file, "r");
	if ( !is_resource($fp) ) {
		echo "Can't open $file\n";
		return 1;
	}
	$stat = fstat($fp);

	while ( !feof($fp) ) {
		$line = trim(ltrim(fgets($fp)));

		if ( empty($line) )
			continue;

		$groupName = NULL;
		$key = NULL;
		$keyName = NULL;
		$keyVal = NULL;

		if ( ereg("^\[(.*)\]$", $line, $groupName) ) {
			$currentGroup = $groupName[1];
		
			/* Replace "/" -> "\/" if its the first char */
/*			if ( $currentGroup[0] == '/' && $currentGroup[1] != '/' )
				$currentGroup = "\/".substr($currentGroup, 1, -1); */

			/* Replace "//" -> "\/\/" */
			$currentGroup = str_replace("/", "\/", $currentGroup);
			continue;
		}

		if ( ($sep = strpos($line, "=")) == FALSE )
			continue;
		$keyName = str_replace("/", "\/", substr($line, 0, $sep));
		$keyVal = substr($line, $sep+1);
	
		$kName = str_replace("\"", "\\\"", "$rootKey$currentGroup/$keyName");
		$kName = str_replace("$", "\\$", $kName);
		
		if ( $keyVal ) {
//			$keyVal = str_replace("\"", "\\\"", $keyVal);
			$keyVal = str_replace("$", "\\$", $keyVal);
			
			
//			echo "kdb set -u ".$stat["uid"]." -g ".$stat["gid"]." -m ".decoct($stat["mode"] & 0777)." -- \"$kName\" \"".str_replace("\"", "\\\"", $keyVal)."\"\n";
			echo "<key uid=\"".$stat["uid"]."\" gid=\"".$stat["gid"]."\" mode=\"".decoct($stat["mode"] & 0777)."\" name=\"$kName\"><value>".str_replace("\"", "\\\"", $keyVal)."</value></key>\n";
			
			
		} else {
//			echo "kdb set -u ".$stat["uid"]." -g ".$stat["gid"]." -m ".decoct($stat["mode"] & 0777)." -- \"$kName\"\n";
			echo "<key uid=\"".$stat["uid"]."\" gid=\"".$stat["gid"]."\" mode=\"".decoct($stat["mode"] & 0777)."\" name=\"$kName\"/>\n";
		}
	}

	
	fclose($fp);
}


function convertDir($path, $rootKey)
{
	$dir = opendir($path);
	if ( !is_resource($dir) ) {
		echo "Can't open $path\n";
		return 1;
	}
	rewinddir($dir);

	while ( ($fname = readdir($dir)) != FALSE ) {
		if ( $fname == "." || $fname == ".." || $fname[0] == '.' )
			continue;

		if ( is_dir("$path/$fname") )
			convertDir("$path$fname/", "$rootKey$fname/");
		else
			convert("$path$fname", "$rootKey$fname/");
	}
}

function getHomesDir()
{
	$fp = fopen("/etc/passwd", "r");
	if ( !is_resource($fp) ) {
		echo "Can't open /etc/passwd\n";
		exit;
	}

	$homes = array( "user"	=> array(),
			"home"	=> array());
	while ( !feof($fp) ) {
		$line = trim(ltrim(fgets($fp)));
		if ( empty($line) )
			continue;

		$data = explode(":", $line);

		if ( ereg("^/home/", $data[5]) ) {
			$homes["user"][] = $data[0];
			$homes["home"][] = $data[5];
		}
	}
	
	fclose($fp);

	return $homes;
}

function getKdePath($homes)
{
	exec("kde-config --path config", $output, $ret);
	$path = explode(":", $output[0]);

	$kdePath = array();
	$kdePath["system"] = $path[1];

	$home = getenv("HOME");
	
	if ( ereg("^$home(.*)$", $path[0], $ret) ) {
		$kdePath["user"] = array();
		for($i = 0 ; $i<count($homes) ; $i++) {
//			if ( file_exists($homes[$i].$ret[1]) )
				$kdePath["user"][] = $homes[$i].$ret[1];
		}	
	} else {
		echo "Can't get the user part of the path. Quit";
		exit;
	}
	
	return $kdePath;
}

// convert("/home/yl/.kde/share/config/cervisiarc", "user/sw/kde/");

$users = getHomesDir();
$kdeConf = getKdePath($users["home"]);

convertDir($kdeConf["system"], "system/sw/kde/");
for($i = 0 ; $i<count($kdeConf["user"]) ; $i++) {
	convertDir($kdeConf["user"][$i], "user:".$users["user"][$i]."/sw/kde/");
}

/* $infile = $argv[1];
$keyroot = $argv[2];

convert($infile, $keyroot); */


?>

File: patches/kconfig/kde2elektra.cpp
/**************************************************************************
 *  Copyright (C) 2006 by Yannick Lecaillez                               *
 *  sizon5@gmail.com                                                      *
 *                                                                        *
 *  This program is free software; you can redistribute it and/or modify  *
 *  it under the terms of the GNU General Public License as published by  *
 *  the Free Software Foundation; either version 2 of the License         *
 *                                                                        *
 *  This program is distributed in the hope that it will be useful,       *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *  GNU General Public License for more details.                          *
 *                                                                        *
 *  You should have received a copy of the GNU General Public License     *
 *  along with this program; if not, write to the                         *
 *  Free Software Foundation, Inc.,                                       *
 *  59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 **************************************************************************/

#include <iostream>
using namespace std;

#include <qdir.h>
#include <kdebug.h>
#include <kapplication.h>
#include <kconfig.h>
#include <kstandarddirs.h>

#include "kdb.h"

int kde2elektra(const QString &absFname, const QString &baseDir)
{
	KeySet		*ks;
	bool		user;
	QString 	relName(absFname);
	QFileInfo	info(absFname);
	
	relName.remove(0, baseDir.length());
	user = baseDir.startsWith(getenv("HOME"));

	ks = ksNew();
	
	KConfig c(absFname);
	QStringList groups = c.groupList();
	for (QStringList::const_iterator it = groups.constBegin();
		         it != groups.constEnd();
		          ++it) {

		QMap<QString, QString> entries = c.entryMap(*it);
		int numEntries = entries.count();

		if (numEntries < 1) {
			continue;
		}

		QString groupName((*it));
		groupName.replace('/', "\\/");

		QMap<QString, QString>::const_iterator endEntries = entries.constEnd();
		for (QMap<QString, QString>::const_iterator entryIt = entries.constBegin();
				entryIt != endEntries;
				++entryIt) {

			if ( entryIt.data().isEmpty() )
				continue;

			QString keyName(entryIt.key());
			keyName.replace('/', "\\/");
						
			QString kName(groupName + "/" + keyName);

			if ( user )
				kName.prepend("user/sw/kde/current/kde/config/" + relName + "/");
			else
				kName.prepend("system/sw/kde/current/kde/config/" + relName + "/");
			
			Key *key;
			
			key = keyNew((const char *) kName.utf8(), 
					KEY_SWITCH_UID, info.ownerId(),
					KEY_SWITCH_GID, info.groupId(),
					KEY_SWITCH_VALUE, (const char *) entryIt.data().utf8(),
					KEY_SWITCH_END);
			ksAppend(ks, key);
		}
	}

	char *oldLocale;
	
	oldLocale = setlocale(LC_ALL, NULL);
	if ( setlocale(LC_ALL, "UTF-8") == NULL ) {
		cout << "Can't change locale to UTF-8" << endl;
		exit(1);
	}

	KDBHandle	handle;
	
	kdbOpen(&handle);
	kdbSetKeys(handle, ks);
	kdbClose(&handle);

	ksDel(ks);

	setlocale(LC_ALL, oldLocale);
	
	return 0;
}
		
int convertDir(const QString &dirName, const QString &baseDir)
{
	QDir 		dir(dirName);
	QStringList	files = dir.entryList();

	for (QStringList::Iterator at = files.begin(); at != files.end(); at++) {
		if ( (*at) == "." || (*at) == ".." )
			continue;
		QString fName(dir.absFilePath((*at), false));
//		fName = dirName;
//		fName.append((*at));
		
		QFileInfo info(fName);
		if ( info.isDir() )
			convertDir(fName, baseDir);

		kde2elektra(fName, baseDir);
	}

	return 0;
}


int main(int argc, char **argv)
{
	QString	dir;
	KApplication  a( argc, argv, "kde2elektra" );
	
	QStringList configDir = KGlobal::dirs()->resourceDirs("config");
	if ( configDir.size() != 2 ) {
		cout << "Your KDE config use more than 2 directories. I can't do nothing for you :-(." << endl;
		exit(1);
	}

	if ( getuid() ) {
		cout << "You're root, importing system wide config." << endl;
		cout << "Re-run me after that as user for importing user specific config (if not done yet)." << endl;
		convertDir(configDir.last(), configDir.last());
		cout << "Importing root specific config ..." << endl;
		convertDir(configDir.first(), configDir.first());
	} else {
		cout << "You're user, importing user specific config." << endl;
		cout << "Re-run me after that as root for importing system wide config (if not done yet)." << endl;
		convertDir(configDir.first(), configDir.first());
	}

	return 0;	
//	kdDebug() << "config dir is " << configDir << endl;
}	
	    
